File: packages/catalog-python/src/ai37_a2ui_catalog/models/simple_table.py
from __future__ import annotations

from pydantic import Field, model_validator

from .shared import CellPrimitive, StrictModel, TextAlign


class SimpleTableColumn(StrictModel):
    key: str = Field(min_length=1)
    header: str = Field(min_length=1)
    align: TextAlign = None
    width: str = Field(default=None, min_length=1)


class SimpleTableRow(StrictModel):
    cells: list[CellPrimitive]


class SimpleTableProps(StrictModel):
    title: str = Field(default=None, min_length=1, max_length=120)
    caption: str = Field(default=None, min_length=1, max_length=240)
    compact: bool = None
    striped: bool = None
    columns: list[SimpleTableColumn] = Field(min_length=1)
    rows: list[SimpleTableRow]

    @model_validator(mode="after")
    def validate_row_lengths(self) -> "SimpleTableProps":
        expected = len(self.columns)
        for index, row in enumerate(self.rows):
            if len(row.cells) != expected:
                raise ValueError(f"Row {index} must contain exactly {expected} cells.")
        return self

File: packages/catalog-schemas/src/components/simple-table.ts
import {z} from 'zod';
import {cellPrimitiveSchema, type CatalogComponentDefinition, textAlignSchema} from '../types';

export const simpleTableColumnSchema = z
  .object({
    key: z.string().min(1),
    header: z.string().min(1),
    align: textAlignSchema.optional(),
    width: z.string().min(1).optional(),
  })
  .strict();

export const simpleTableRowSchema = z
  .object({
    cells: z.array(cellPrimitiveSchema),
  })
  .strict();

export const simpleTablePropsSchema = z
  .object({
    title: z.string().min(1).max(120).optional(),
    caption: z.string().min(1).max(240).optional(),
    compact: z.boolean().optional(),
    striped: z.boolean().optional(),
    columns: z.array(simpleTableColumnSchema).min(1),
    rows: z.array(simpleTableRowSchema),
  })
  .strict()
  .superRefine((value, ctx) => {
    for (const [rowIndex, row] of value.rows.entries()) {
      if (row.cells.length !== value.columns.length) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: `Row ${rowIndex} must contain exactly ${value.columns.length} cells.`,
          path: ['rows', rowIndex, 'cells'],
        });
      }
    }
  });

export type SimpleTableProps = z.infer<typeof simpleTablePropsSchema>;

export const simpleTableDefinition: CatalogComponentDefinition<typeof simpleTablePropsSchema> = {
  name: 'SimpleTable',
  slug: 'simple-table',
  description:
    'A simple data table with a single header row. Use it for straightforward tabular datasets where each row follows the same column structure.',
  schema: simpleTablePropsSchema,
};
